src/DocumentFormat.OpenXml/ParaId/Paragraph.ParaId.cs

```diff
@@ -0,0 +1,128 @@
+        private protected override void SetAttribute<TSimpleType>(TSimpleType? value, [CallerMemberName] string propertyName = null!)
+            where TSimpleType : class
+        {
+            if (propertyName is nameof(ParagraphId) && value is HexBinaryValue hexBinaryValue)
+            {
+                // TODO: Discuss how we want to deal with duplicate w14:paraId values assigned by the caller.
+                // The RegisterParagraphId() method returns true if the hexBinaryValue was unique and false otherwise.
+                // Therefore, if desired (e.g., always or based on settings), we could throw an exception.
+                WordprocessingDocument?.RegisterParagraphId(hexBinaryValue);
+            }
+
+            base.SetAttribute(value, propertyName);
+        }
+
+        /// <summary>
+        /// Attempts to set the attribute to a known attribute and registers any existing
+        /// w14:paraId values with the <see cref="WordprocessingDocument" /> instance.
+        /// </summary>
+        /// <remarks>
+        /// This is the Paragraph-specific implementation that registers the w14:paraId
+        /// values when loading the DOM tree.
+        /// </remarks>
+        /// <param name="qname">The attribute's qualified name.</param>
+        /// <param name="value">The attribute's value.</param>
+        /// <param name="strictRelationshipFound"></param>
+        /// <returns>true if the attribute is a known attribute.</returns>
+        private protected override bool TrySetFixedAttribute(in OpenXmlQualifiedName qname, string? value, bool strictRelationshipFound)
+        {
+            if (RawState.Attributes.Any())
+            {
+                if (strictRelationshipFound)
+                {
+                    return StrictTranslateAttribute(qname, value);
+                }
+
+                AttributeCollection.AttributeEntry attribute = RawState.Attributes[qname];
+
+                if (!attribute.IsNil)
+                {
+                    attribute.Value ??= attribute.Property.CreateNew();
+                    attribute.Value.InnerText = value;
+
+                    if (qname.Name == "paraId" && value is not null)
+                    {
+                        WordprocessingDocument?.RegisterParagraphId(value);
+                    }
+
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <inheritdoc />
+        public string SetUniqueParagraphId()
+        {
+            WordprocessingDocument? wordprocessingDocument = WordprocessingDocument;
+            if (wordprocessingDocument is null)
+            {
+                throw new InvalidOperationException("Paragraph must be added to the DOM tree before calling this method.");
+            }
+
+            return SetUniqueParagraphId(wordprocessingDocument);
+        }
+
+        /// <inheritdoc />
+        public string SetUniqueParagraphId(IParagraphIdService paragraphIdService)
```

Comment:
let's make this internal for now - I've got some ideas I'll play around with how to make it configureable for the end user

Reply:
This method would be useful in use cases where you would like to provide your own `IParagraphIdService` instance when setting those values. For example, I have a specific use case in which I would like to ensure the uniqueness of `w14:paraId` values across a set of `WordprocessingDocument` instances rather than "just" within an individual `WordprocessingDocument`.

What kind of configuration are you thinking about?

Reply:
I agree with the purpose of it, but I'm not happy with the shape of the API at the moment.

I've been toying with some ideas of services and how a user might want to replace/configure things. ie have a basic IOC container at the root (such as the OpenXmlPartContainer) that exposes `IServiceProvider` so we can provide services without coupling ourselves to specific implementations. This fits in with that goal and would like to explore a more generalized infrastructure. Keeping it internal allows us to play with that without having to decide on that shape right now.